### DISCUSSION TITLE
index column

### DISCUSSION THREAD

**Discussion by Ajayghoshm** [TanStack/table#2224] [General] [answered]
I need to add a index column as the first column, which doesn't get effected by the sorting,
currently i added a index column and index data as the props passed,
but this column will get sorted (change the index order, becomes random ordering number)
is there any way to solve the issue

Before Sorting

![Screenshot 2020-04-25 at 9 42 36 AM](https://user-images.githubusercontent.com/29367198/80270851-241b1000-86d9-11ea-83b2-cd6fb283d117.png)

After Sorting

![Screenshot 2020-04-25 at 9 42 29 AM](https://user-images.githubusercontent.com/29367198/80270849-22514c80-86d9-11ea-8fee-2afa4bfcfbd8.png)

**Comment by IamDhana** [comment]
Render index column as custom cell 
[https://codesandbox.io/s/wonderful-wozniak-sn3yt?file=/src/App.js](url)

**Reply by Ajayghoshm** [reply]
This one still get sorting issue

**Reply by Ajayghoshm** [reply]
` hooks => {
      hooks.visibleColumns.push(columns => [
        {
          id: "count",
          isIndex: true,
          width: 50,
          Header: 'No',
          disableFilters: true,
          disableSortBy: true,
          Cell: ({ row }) => {
            console.log("no",row)
            return(
            <Fragment>
              {row.index+1}
            </Fragment>
            )
            }
        },
        ...columns
      ]);
    }`

i tried this approach , still no solved

**Comment by gargroh** [comment]
Bad approach - but what upfront is coming in my mind -

Try to check on `instance.rows` i.e. in `Cell` try to find the `index` of given `row` and return its index.

**Reply by IamDhana** [reply]
May I know why it is a bad approach?

**Reply by Ajayghoshm** [reply]
If you are talking about the index value inside the "Cell"
`Cell: ({ row }) => {
            console.log("no",row,row.index)
            return(
            <Fragment>
              {row.index}
            </Fragment>
            )
            }`
this index value changes order on sorting.

**Comment by Ajayghoshm** [comment, marked as answer]
I was able to solve this issue, defined a actual row inside the actual table component which count the row by using the index value in the map function of row

**Reply by Multisanti64** [reply]
I'm so glad i found your answer. I've had the same problem. I followed your way and solved it like this:
Added indexed attribute to column:
`const columns = useMemo<Column<typeof result[number]>[]>(() => [
        {
            Header: '№',
            id: 'index',
            disableSortBy: true,
            indexed: true, <-- custom attribute
        }
]`
And then checked this property in render:
`
page.map((row, i) => {
    prepareRow(row)
    return (
        <tr {...row.getRowProps()}>
            {
                row.cells.map((cell) => (
                    <td {...cell.getCellProps()}
                        className={cell.column.classes ? cell.column.classes : ''}>
                        {cell.column.indexed ? i + 1 : cell.render('Cell')}
                    </td>
                ))
            }
        </tr>
    )
})
`

**Comment by gsh4rm4** [comment]
[viewIndex was accessible in V6](https://github.com/tannerlinsley/react-table/issues/786#issuecomment-362878460), which solved the challenge listed in this discussion.

It looks like V7 removed `viewIndex`. I'm not sure if @gary-menzel or someone else knows how to access `viewIndex` of a row in V7?

**Comment by henry61024** [comment]
You could use `Cell` render props' `row` and `flatRow` arguments to achieve this requirement

```jsx
// columns
[
  {
    Header: 'No',
    id: 'id',
    Cell: ({ row, flatRows }) => {
      return flatRows.indexOf(row) + 1;
    },
  },
  // ...other columns
]
```

`Cell` render props document: https://react-table.tanstack.com/docs/api/useTable#cell-properties

**Comment by changyeamoon** [comment]
@henry61024 any idea on how to do this with Tanstack table 8? Dont see flatRows exposed anymore

**Reply by HareeshJakkampudi** [reply]
@changyeamoon 

try this 
example:
cell:(props)=>{
     
     return props?.table?.getSortedRowModel()?.flatRows?.indexOf(props?.row)+1;
}

**Reply by YuriyDyachkov** [reply]
> @changyeamoon
> 
> try this example: cell:(props)=>{
> 
> ```
>  return props?.table?.getSortedRowModel()?.flatRows?.indexOf(props?.row)+1;
> ```
> 
> }

Please tell me how can i do this with expanded rows?

**Comment by lazy-ocean** [comment]
I had to modify pieces from above comments to make it work with sorting applied, maybe this will help someone as well:
```
{
  Header: "No",
  id: "id",
  cell: ({ row, table }) =>
         (table.getSortedRowModel()?.flatRows?.findIndex((flatRow) => flatRow.id === row.id) || 0) + 1,
}
```

**Reply by sghong3130** [reply]
사랑해요

**Reply by Zain-ul-din** [reply]
Thanks to @lazy-ocean 

we can also use this for (shift + click) multi-select:

![image](https://github.com/user-attachments/assets/ffbe5b29-c9b7-4c6b-a2a2-b4f24ce019b0)
 
```ts
 let lastSelectedID = null
```
 
 ```ts
 cell: ({ row, table }) => {
        const nonSortedIdx =
          table.getSortedRowModel()?.flatRows?.findIndex((flatRow) => flatRow.id === row.id) || 0

        return (
          <div className="flex justify-center items-center w-full pr-4">
            <p>
              {nonSortedIdx + ' ' + row.id}{' '}
              <Checkbox
                checked={row.getIsSelected()}
                onClick={(e) => {
                  if (e.shiftKey) {
                    const { rows, rowsById } = table.getRowModel()
                    const rowsToToggle = getRowRange(
                      rows,
                      Number(nonSortedIdx),
                      Number(lastSelectedID)
                    )
                    const isCellSelected = rowsById[nonSortedIdx].getIsSelected()
                    rowsToToggle.forEach((_row) => _row.toggleSelected(!isCellSelected))
                  } else {
                    row.toggleSelected()
                  }
                  lastSelectedID = nonSortedIdx
                }}
                onCheckedChange={(value) => {
                  row.toggleSelected(!!value)
                }}
                aria-label="Select row"
              />
            </p>
          </div>
        )
      },
      enableSorting: true,
      enableHiding: true
    },
    
   ```
   
   _util_
   
   ```ts
   export const getRowRange = (
  rows: Row<any>[],
  currentID: number,
  selectedID: number
): Row<any>[] => {
   const rangeStart = selectedID > currentID ? currentID : selectedID
   const rangeEnd = rangeStart === currentID ? selectedID : currentID
   return rows.slice(rangeStart, rangeEnd + 1)
}
```

**Comment by rajagopalstalwart** [comment]
@lazy-ocean its working Good, Thank you for providing

**Comment by therealokoro** [comment]
From the future ...

April 11, 2024. @lazy-ocean answer works perfectly

**Comment by aboveyunhai** [comment]
@lazy-ocean solution works perfectly, 
but I doubt that the `rows.findIndex` might cause `O(n^2)` performance when you have thousands of rows and do index find for each cell. 
That could be terribly slow without knowing how the cell function is handled internally,

Therefore, my idea is built upon @lazy-ocean solution, just cache the array in map upon the sorting change, and do the idx on rendering.
Simply convert the  `O(n^2)` to `O(n)`
``` typescript
  const sortedRows = table.getSortedRowModel().flatRows;
   // this will only run upon sorting change.
  const sortedMap = useMemo(() => {
    const output: Record<string, number> = {};
    sortedRows.forEach((row, idx) => {
      output[row.id] = idx;
    });
    return output;
  }, [sortedRows]);
```
``` jsx

 <div>
    {cell.column.id === "id"
       ? sortedMap[row.id] + 1
       : flexRender( cell.column.columnDef.cell, cell.getContext())
    }
</div>
```

**Comment by johnny-mh** [comment]
Using content selector (it also work with virtualRow.index on @tanstack/virtual)

```jsx
<tbody>
  {table.getRowModel().rows.map((row, idx) => {
    return <tr className="first-of-type:[&>td]:before:content-[attr(data-rownum)]">
      {row.getVisibleCells().map(cell => {
        return <td data-rownum={idx+1}></td>
      })}
    </tr>
  })}
</tbody>
